File: fsharp/grade-school/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = Map.empty

let add (student: string) (grade: int) (school: School): School =
    let students = 
        school
        |> Map.tryFind grade
        |> Option.map (fun students -> student :: students)
        |> Option.defaultValue [student]
    school
    |> Map.add grade students

let roster (school: School): string list =
    school
    |> Map.toSeq
    |> Seq.sortBy fst
    |> Seq.map (snd >> List.sort) 
    |> List.concat

let grade (number: int) (school: School): string list =
    school
    |> Map.tryFind number
    |> Option.map List.sort
    |> Option.defaultValue []

File: javascript/grade-school/grade-school.js
export class GradeSchool {
  constructor() {
    this._roster = {};
  }

  roster() {
    return JSON.parse(JSON.stringify(this._roster));
  }

  add(student, grade) {
    removeFromRoster(this._roster, student);
    addToRoster(this._roster, student, grade);
  }

  grade(number) {
    return [...(this._roster[number] || [])];
  }
}

const removeFromRoster = (roster, student) => {
  for (const grade of Object.values(roster)) {
    const index = grade.indexOf(student);
    if (index !== -1) grade.splice(index, 1);
  }
};

const addToRoster = (roster, student, grade) => {
  const students = roster[grade];
  roster[grade] = [...(students || []), student].sort();
};
